### PR TITLE
Update usage of unsafe API, upgrade libarchive to 3.7.4, upgrade async http client

### PR DESCRIPTION
**Pull request by cmcgee1024**
Reduce the use of unsafe swift constructs in favour of ones that are memory safe. In the HTTP client remove the use of the unsafe mutable pointer in favour of direct Data conversion before writing the downloaded file to disk.

Upgrade the version of libarchive to 3.7.4 with various fixes.

Upgrade the async http client to fix an issue with decompression limits encountered with the GitHub REST API's.

### CODE REVIEW

File: Sources/SwiftlyCore/HTTPClient.swift
Comment (adam-fowler) [inline Sources/SwiftlyCore/HTTPClient.swift:171]: Can probably replace all of this with 
```swift
try fileHandle.write(contentsOf: buffer.readableBytesView)
```

**Review by adam-fowler — APPROVED** [review]
Hi Chris, welcome onboard
This looks good, but I've a suggested an alternative.

File: Sources/SwiftlyCore/HTTPClient.swift
Comment (cmcgee1024) [reply Sources/SwiftlyCore/HTTPClient.swift]: Thank you for the suggestion, this is much cleaner. I've pushed an update that uses it.

File: Sources/SwiftlyCore/HTTPClient.swift
Comment (cmcgee1024) [inline Sources/SwiftlyCore/HTTPClient.swift:186]: Note that this was added because in ad hoc testing I discovered an emergent http decompression limit error. It appears that something has changed to trigger it in the GitHub API's.

Comment (adam-fowler) [reply Sources/SwiftlyCore/HTTPClient.swift:186]: Does `.shared` have different limits to the HTTPClient with a default configuration?

Comment (cmcgee1024) [reply Sources/SwiftlyCore/HTTPClient.swift:186]: It appears so. I dug a little deeper because I was concerned that this could be affecting everyone using swiftly. This is the commit that regressed and the testing didn't pick it up at the time: 2d054143ea29e2e5ee0085cbb0ca2ed8086d9166

I'm fully reverting that change so that this is not broken, but it does leave some open questions about our commit verification process. Ideally, some form integration testing might have picked this up at the time. Also, it does seem suspect that the shared HTTP Client has a broken decompression limit, whereas this initializer that just reuses the singleton event loop group provider has an intact decompression.

Comment (adam-fowler) [reply Sources/SwiftlyCore/HTTPClient.swift:186]: Have you reported this in the async-http-client repository?

Comment (adam-fowler) [reply Sources/SwiftlyCore/HTTPClient.swift:186]: Including @weissi as I believe this was from his PR to use shared

Comment (cmcgee1024) [reply Sources/SwiftlyCore/HTTPClient.swift:186]: I can try and create a small reproducible test case and see if the latest fixes this. I hope to do that once I've restored this to a working state here with the original working code.

Comment (cmcgee1024) [reply Sources/SwiftlyCore/HTTPClient.swift:186]: @adam-fowler the latest commit fully reverts the HTTPClient.swift portion of 2d0514 and smoke testing seems to be working again. Any chance that we can get these fixes in to get us back into a working state?

Comment (gwynne) [reply Sources/SwiftlyCore/HTTPClient.swift:186]: We've run into this problem over in Penny (Vapor's Discord bot) - this was the fix Mahdi implemented: https://github.com/vapor/penny-bot/blob/main/Sources/Shared/%2BHTTPClient.swift#L12-L14

Comment (MahdiBM) [reply Sources/SwiftlyCore/HTTPClient.swift:186]: Yes, and there is already an issue for it (which I should have marked as completed, but forgot):
https://github.com/swift-server/async-http-client/issues/739

@cmcgee1024 feel free to comment there if you think the limit is still too low for Swiftly, although I'm not sure how much success you'll have with getting the team to agree on increasing it, considering increasing the limit has some implications.

Comment (cmcgee1024) [reply Sources/SwiftlyCore/HTTPClient.swift:186]: Thanks @gwynne 

I think that unless there's a reason why we can't revert the HTTPClient change here, we should probably go with that for now until this issue is fixed in async-http-client: https://github.com/swift-server/async-http-client/issues/739

Comment (adam-fowler) [reply Sources/SwiftlyCore/HTTPClient.swift:186]: Having investigated, the reason `HTTPClient(eventLoopGroupProvider: .singleton)` works and `HTTPClient.shared` doesn't is the first disables decompression. I would rather we still ran with decompression enabled. If the current rate is too low then let's do what @MahdiBM suggests. I would also like this implemented as a singleton so we don't have to worry about shutdown.

Comment (Joannis) [reply Sources/SwiftlyCore/HTTPClient.swift:186]: @adam-fowler heads up: you've approved this PR, but left a change request in this comment afterwards.

Comment (patrickfreed) [reply Sources/SwiftlyCore/HTTPClient.swift:186]: >Ideally, some form integration testing might have picked this up at the time.

It appears maybe this was some change on the GitHub side that introduced this issue. I've just tried to run a basic install locally and hit the error, so I'd imagine if we had CI set up right now we'd see it. But yeah, it does indicate that we should be running tests on some regular interval (daily?) to ensure things like this can be caught as they happen.

**Comment by patrickfreed** [comment]
I just looked into the background for the libarchive security issue, and in some ways it's pretty serious as it's related to the known bad actor that compromised the xz library. See https://github.com/libarchive/libarchive/pull/2101 and https://github.com/libarchive/libarchive/pull/1609 for more context.

That being said, it looks like no CVE was issued for it, and the suspicious code changes only affected the bsdtar utility, rather than the library functions as used by swiftly. So in light of that, I don't believe there are security concerns here for us. We should still proceed with the upgrade, but I don't think we should indicate that doing so is a security fix in the commit message or release notes, as that has certain implications for downstream users.

Likewise, I agree that moving away from using unsafe APIs in swiftly is an improvement that we should merge, but the case identified here was a safe usage of it, so there was no memory safety violation that we're now fixing. In the commit message / release notes, we should note this accordingly (e.g. "stop using unsafe API" rather than "memory safety fixes"). Again, the only reason I'm harping on this is security issues have to be handled with special care, and so if we did have an identified security problem, we would want to cut a release as soon as possible and notify users in the changelog what the case we're fixing is. Thankfully, that doesn't appear to be the case here though.

All that being said, changes look good to me mod @adam-fowler's comment about using a singleton.

**Comment by patrickfreed** [comment]
I took the liberty of updating the PR title, let me know if it seems alright with you.

File: Sources/SwiftlyCore/HTTPClient.swift
Comment (FranzBusch) [reply Sources/SwiftlyCore/HTTPClient.swift:186]: I agree with @adam-fowler here let's define a static one so we don't need the wrapper with the `deinit`

Comment (cmcgee1024) [reply Sources/SwiftlyCore/HTTPClient.swift:186]: > It appears maybe this was some change on the GitHub side that introduced this issue. I've just tried to run a basic install locally and hit the error, so I'd imagine if we had CI set up right now we'd see it. But yeah, it does indicate that we should be running tests on some regular interval (daily?) to ensure things like this can be caught as they happen.

After thinking on this a bit, I don't think that GitHub is the variable as I had initially thought. Everything was working for me until I rebased this change on top of the above commit that changed the way that HTTPClient was initialized. A CI with at least a smoke test that did a `swiftly install latest` would have picked up the problem.

Comment (cmcgee1024) [reply Sources/SwiftlyCore/HTTPClient.swift:186]: >  I would rather we still ran with decompression enabled. If the current rate is too low then let's create a client with a higher rate. I would also like this implemented as a singleton so we don't have to worry about shutdown. eg

Is this blocking the approval and merge (I don't merge permissions here) of this PR? Can we defer this to a new issue/PR? It appears that the existing approach before the regression was working for quite some time and is currently working for everyone on v0.3.0. I'm just restoring that so that we're all unblocked.

Comment (adam-fowler) [reply Sources/SwiftlyCore/HTTPClient.swift:186]: > Is this blocking the approval and merge (I don't merge permissions here) of this PR? Can we defer this to a new issue/PR?

I would rather we stick to a singleton. The change I suggested works, as I used it when testing the macOS support PR

Comment (weissi) [reply Sources/SwiftlyCore/HTTPClient.swift:186]: When you say 'a singleton', you mean `HTTPClient.shared` or your own singleton?

I very much think it should be `HTTPClient.shared`, there's no reason to create singletons all over the place, just wastes resources. The design goal is that `HTTPClient.shared` behaves like the platform browser: If Safari/Chrome/... can open it in defaults settings then `HTTPClient.shared` should do.

Comment (adam-fowler) [reply Sources/SwiftlyCore/HTTPClient.swift:186]: @weissi I understand that but we cannot use `HTTPClient.shared` at the moment. It just keeps throwing errors about compression limits.

Comment (cmcgee1024) [reply Sources/SwiftlyCore/HTTPClient.swift]: @adam-fowler I've updated this PR with your suggested change to the construction of the http client.

Comment (weissi) [reply Sources/SwiftlyCore/HTTPClient.swift]: @adam-fowler Correct right this instant but the fix for this is merged, I'm working with the maintainers on cutting a release

Comment (weissi) [reply Sources/SwiftlyCore/HTTPClient.swift]: Thanks to @lukasa, we've got https://github.com/swift-server/async-http-client/releases/tag/1.21.2 now. So `from: "1.21.2"` for AHC should fix this.

Comment (patrickfreed) [reply Sources/SwiftlyCore/HTTPClient.swift]: @weissi / @Lukasa thanks for pushing that through!

>After thinking on this a bit, I don't think that GitHub is the variable as I had initially thought. Everything was working for me until I rebased this change on top of the above commit that changed the way that HTTPClient was initialized. A CI with at least a smoke test that did a swiftly install latest would have picked up the problem.

Prior to the move to swiftlang, we did run tests in CI, a few of which covered that case (`testInstallLatest` and `default-install.sh` in particular). That's what leads me to believe maybe something changed on the GitHub side. @shahmishal is working on getting CI set up under the new organization, so we'll have automated testing back shortly.

**Review by patrickfreed — APPROVED** [review]

**Comment by shahmishal** [comment]
@swift-server-bot test install please

**Comment by shahmishal** [comment]
@swift-server-bot test this please

**Comment by shahmishal** [comment]
@swift-server-bot test install please

**Comment by shahmishal** [comment]
@swift-server-bot test install please

**Comment by shahmishal** [comment]
@swift-server-bot test this please

**Comment by cmcgee1024** [comment]
Thanks everyone for your help with this.

**Pull request merged**
